### DISCUSSION TITLE
How can I customize SWC configuration to target ES2015+?

### DISCUSSION THREAD

**Discussion by hueter** [shakacode/shakapacker#342] [Q&A] [answered]
Hello,

I am using SWC with Shakapacker and it's great! 🎉 

However, I ran into an issue today. I'm trying to transpile to ES2015+ so I can use [web components in my app](https://stackoverflow.com/questions/39037489/extending-htmlelement-constructor-fails-when-webpack-was-used), so I went to the SWC config and tried to set `target: "es2015"`.

However, this setting is incompatible with Shakapacker's default SWC `env` setting, defined here: https://github.com/shakacode/shakapacker/blob/cdf32835d3e0949952b8b4b53063807f714f9b24/package/swc/index.js#L37-L40

When I specify [`target` as per the SWC docs](https://swc.rs/docs/configuration/compilation#jsctarget), it results in a compilation error:

<img width="744" alt="Screenshot 2023-08-02 at 10 31 10 PM" src="https://github.com/shakacode/shakapacker/assets/13444851/7aa6aa28-5992-4bcf-b595-b41963498249">

I went poking around `node_modules` to see if I could fix it, and this worked:

```tsx
module.exports = loaderMatches(webpackLoader, 'swc', () => ({
  test: /\.(ts|tsx|js|jsx|mjs|coffee)?(\.erb)?$/,
  ...jscommon,
  use: ({ resource }) => {
    const swcConfig = getSwcLoaderConfig(resource);
    // I added this line below
    delete swcConfig.options.env
    return swcConfig
  }
}))
```

But as you can tell, this is janky and I would rather not fork the repo.

Do y'all have any recommendations? Or an alternative strategy to transpile to ES2015+?

Thank you!

**Comment by ahangarha** [comment]
Thanks for giving us feedback.

CC @justin808 @Judahmeek

**Comment by hueter** [comment, marked as answer]
Update: I actually found a workaround that transpiles to ES2015+, just by setting the [browserslist config](https://github.com/browserslist/browserslist#full-list) in the [`env.targets`](https://swc.rs/docs/configuration/compilation#envtargets):

`swc.config.js`

```js
const customConfig = {
  options: {
    jsc: {/* etc */},
    env: {
      // this overrides the Shakapacker defaults
      coreJs: null,
      exclude: [],
      mode: "entry",
      // this transpiles to ES2021+
      targets: "last 2 years",
    },
  },
};

module.exports = customConfig;
```

**Reply by ahangarha** [reply]
So basically you needed to override the default config for `env` entries.

Would be nice to include in the documentation.
